zebra-chain/src/transaction/unmined/zip317.rs

```diff
@@ -139,3 +147,64 @@ fn conventional_actions(transaction: &Transaction) -> u32 {
 
     max(GRACE_ACTIONS, logical_actions)
 }
+
+/// Make ZIP-317 checks before inserting a transaction into the mempool.
+pub fn mempool_checks(
+    unpaid_actions: u32,
+    miner_fee: Amount<NonNegative>,
+    conventional_fee: Amount<NonNegative>,
+    transaction_size: usize,
+) -> Result<(), Error> {
+    // # Standard Rule
+    //
+    // > If a transaction has more than `block_unpaid_action_limit` "unpaid actions" as defined by the
+    // > Recommended algorithm for block template construction, it will never be mined by that algorithm.
+    // > Nodes MAY drop these transactions.
+    //
+    // <https://zips.z.cash/zip-0317#transaction-relaying>
+    if unpaid_actions > BLOCK_PRODUCTION_UNPAID_ACTION_LIMIT {
+        return Err(Error::UnpaidActions);
+    }
+
+    // # Standard Rule
+    //
+    // > Nodes that normally relay transactions are expected to do so for transactions that pay at least the
+    // > conventional fee as specified in this ZIP.
+    //
+    // <https://zips.z.cash/zip-0317#transaction-relaying>
+    if miner_fee < conventional_fee {
+        return Err(Error::FeeBelowConventional);
+    }
+
+    // # Standard rule
+    //
+    // > Minimum Fee Rate
+    // >
+    // > Transactions must pay a fee of at least 100 zatoshis per 1000 bytes of serialized size,
+    // > with a maximum fee of 1000 zatoshis. In zcashd this is `DEFAULT_MIN_RELAY_TX_FEE`.
+    //
+    // <https://github.com/ZcashFoundation/zebra/issues/5336#issuecomment-1506748801>
+    let limit = std::cmp::min(
+        1000,
+        (f32::floor(100.0 * transaction_size as f32 / 1000.0)) as u64,
+    );
+    if miner_fee < Amount::<NonNegative>::try_from(limit).expect("limit value is invalid") {
```

Comment:
This can probably be improved, i just added [Daira's formula](https://github.com/ZcashFoundation/zebra/issues/5336#issuecomment-1506748801) as it is posted but it needs more work. Any hint is appreciated.

Reply:
I'm going to need the compiler to help me with this one, so I'll open another PR on top of this PR.

Reply:
https://github.com/ZcashFoundation/zebra/pull/6585/files